cmd/opentelemetry-prometheus-sidecar/main.go

```diff
@@ -237,52 +295,28 @@ func Main() bool {
 	level.Debug(logger).Log("msg", "entering run state")
 	healthChecker.SetRunning()
 
-	// Run two inter-depdendent components:
-	// (1) Prometheus reader
-	// (2) Queue manager
-	// TODO: Replace this with x/sync/errgroup
-	var g run.Group
-	{
-		g.Add(
-			func() error {
-				level.Info(logger).Log("msg", "starting Prometheus reader")
-				err = prometheusReader.Run(ctx, startOffset)
-				return err
-			},
-			func(err error) {
-				// Prometheus reader needs to be stopped before closing the TSDB
-				// so that it doesn't try to write samples to a closed storage.
-				// See the use of `stopCh` below to explain how this works.
-				level.Info(logger).Log("msg", "stopping Prometheus reader")
-				cancelMain()
-			},
-		)
-	}
-	{
-		stopCh := make(chan struct{})
-		g.Add(
-			func() error {
-				if err := queueManager.Start(); err != nil {
-					return err
+	g := getRunGroup(ctx, cancelMain, logger, startOffset, prometheusReader, queueManager)
+
+	for {
+		if err := g.Run(); err != nil {
+			if err == tail.ErrRestartReader {
+				level.Error(logger).Log("msg", "restarting reader", "err", err)
+				startOffset, err = readWriteStartOffset(cfg, logger)
+				if err != nil {
+					return false
 				}
-				level.Info(logger).Log("msg", "starting OpenTelemetry writer")
-				<-stopCh
-				return nil
-			},
-			func(err error) {
-				if err := queueManager.Stop(); err != nil {
-					level.Error(logger).Log(
-						"msg", "stopping OpenTelemetry writer",
-						"err", err,
-					)
+				prometheusReader, queueManager, err = initReader(ctx, logger, cfg, scf, promURL, metadataCache, metricRenames, svcInstanceId)
+				if err != nil {
+					return false
 				}
-				close(stopCh)
-			},
-		)
-	}
-	if err := g.Run(); err != nil {
-		level.Error(logger).Log("msg", "run loop error", "err", err)
-		return false
+				// TODO: determine if we should add a max number of retry here
```

Comment:
The `segmentErrorCounter` in `tail.go` counts one of the restarts, but not the existing use of `ErrRestartReader`. Suggest moving the new metric here.